### PR TITLE
Feature/s4

### PR DESCRIPTION
**Pull request by imgeorgiev**
Mostly creating this for easy commenting and keeping track of progress

### CODE REVIEW

File: tdmpc2/tdmpc2.py [hunk 1]
```diff
@@ -299,30 +303,48 @@ def update(self, buffer):
...
+        # z = self.model.encode(obs[0], task)
+        # zs[0] = z
+        # consistency_loss = 0
+        # for t in range(self.cfg.horizon):
+        #     z = self.model.next(z, action[t], task)
+        #     consistency_loss += F.mse_loss(z, next_z[t]) * self.cfg.rho**t
+        #     zs[t + 1] = z
         # Compute targets
         with torch.no_grad():
+            zs = self.model.encode(obs, task)
```
Comment (imgeorgiev) [inline tdmpc2/tdmpc2.py]: I suspect this line here is your culprit. Because its wrapped in a `torch.no_grad()` it means that your line is equivalent to `zs = self.model.encode(obs,task).detach()`. In other words, your encoder is not getting trained since no gradients flow through it anymore. 

Try putting this outside the `with`

File: tdmpc2/tdmpc2.py [hunk 2]
```diff
@@ -299,30 +303,48 @@ def update(self, buffer):
...
-        )
-        z = self.model.encode(obs[0], task)
-        zs[0] = z
         consistency_loss = 0
         for t in range(self.cfg.horizon):
-            z = self.model.next(z, action[t], task)
-            consistency_loss += F.mse_loss(z, next_z[t]) * self.cfg.rho**t
-            zs[t + 1] = z
-
+            consistency_loss += F.mse_loss(zs_hat[t], zs[t+1]) * self.cfg.rho**t
```
Comment (imgeorgiev) [inline tdmpc2/tdmpc2.py]: This here should probably be `consistency_loss += F.mse_loss(zs_hat[t], naxt_z[t]) * self.cfg.rho**t` since you want to compare to detached targets.

File: tdmpc2/tdmpc2.py [hunk 3]
```diff
@@ -299,30 +303,48 @@ def update(self, buffer):
...
+        dynamics_input = torch.cat([dynamics_input, action], dim=-1) # concatenate action to dynamics_input
+        dynamics_input = dynamics_input.permute(1, 2, 0) # Reshape zs from (L, B, H) -> (B, H, L) for dynamics model input 
+        # Encode history by S4 block
+        hs, s_T = self.model._dynamics(dynamics_input)
 
-        # Prepare for update
-        self.optim.zero_grad(set_to_none=True)
-        self.model.train()
+        mlp_input = hs.permute(2, 0, 1)
+        zs_hat = self.model._mlp(mlp_input)
```
Comment (imgeorgiev) [inline tdmpc2/tdmpc2.py]: I need to read up more on S4 models but this here seems weird. No action item yet, just leaving this note for me to check afterwards.

Comment (varungiridhar) [reply tdmpc2/tdmpc2.py]: I tried to format input based on S4Block forward function (https://github.com/state-spaces/s4/blob/e757cef57d89e448c413de7325ed5601aceaac13/models/s4/s4.py#L1892-L1898) which is commonly used in "convolution mode" for training.
